### PR TITLE
Add GetDistributedObjectsInfo API.

### PR DESCRIPTION
**Pull request by enozcan**
Replaces #602 

- Moves DistributedObjectInfo from internal pkg to types. It's also the return type of the new API.
- Unlike #602, this does not create local proxies for fetched distributed objects until client makes `Get*` call.

### CODE REVIEW

File: client.go [hunk 1]
```diff
@@ -200,6 +201,42 @@ func (c *Client) GetPNCounter(ctx context.Context, name string) (*PNCounter, err
...
+	objects := codec.DecodeClientGetDistributedObjectsResponse(resp)
+	result := make([]types.DistributedObjectInfo, 0, len(objects))
+
+	cached := make(map[types.DistributedObjectInfo]struct{})
+	for _, o := range c.proxyManager.getCachedObjectsInfo() {
+		cached[o] = struct{}{}
+	}
+
+	for _, o := range objects {
+		if !serviceSupported(o.ServiceName()) {
```
Comment (yuce) [inline client.go:225]: There's no need to check whether the service is supported, let's return everything.

File: client.go [hunk 2]
```diff
@@ -200,6 +201,42 @@ func (c *Client) GetPNCounter(ctx context.Context, name string) (*PNCounter, err
...
+		return nil, err
+	}
+
+	objects := codec.DecodeClientGetDistributedObjectsResponse(resp)
+	result := make([]types.DistributedObjectInfo, 0, len(objects))
+
+	cached := make(map[types.DistributedObjectInfo]struct{})
+	for _, o := range c.proxyManager.getCachedObjectsInfo() {
+		cached[o] = struct{}{}
+	}
```
Comment (yuce) [inline client.go:222]: I think we shouldn't do anything with the proxy cache anymore.

File: client.go [hunk 3]
```diff
@@ -200,6 +201,42 @@ func (c *Client) GetPNCounter(ctx context.Context, name string) (*PNCounter, err
...
+
+	for _, o := range objects {
+		if !serviceSupported(o.ServiceName()) {
+			continue
+		}
+		result = append(result, o)
+		delete(cached, o)
```
Comment (yuce) [inline client.go:229]: I think we shouldn't do anything with the proxy cache anymore.

File: client.go [hunk 4]
```diff
@@ -200,6 +201,42 @@ func (c *Client) GetPNCounter(ctx context.Context, name string) (*PNCounter, err
...
+	}
+
+	for o := range cached {
+		// Purge the stale object from the manager's cache.
+		c.proxyManager.remove(o.ServiceName(), o.Name())
+	}
```
Comment (yuce) [inline client.go:235]: I think we shouldn't do anything with the proxy cache anymore.

File: internal/it/util.go [hunk 5]
```diff
@@ -190,6 +190,10 @@ func MustClient(client *hz.Client, err error) *hz.Client {
 	return client
 }
 
+func NewUniqueServiceName(service string) string {
```
Comment (yuce) [inline internal/it/util.go:193]: This function seems to be creating object names, can you rename it to `NewObjectName`?

File: types/distributed_objects_info.go [hunk 6]
```diff
@@ -0,0 +1,34 @@
...
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package types
+
+type DistributedObjectInfo struct {
+	name        string
+	serviceName string
```
Comment (yuce) [inline types/distributed_objects_info.go:21]: Let's export these fields.

File: types/distributed_objects_info.go [hunk 7]
```diff
@@ -0,0 +1,34 @@
...
+}
+
+func (i *DistributedObjectInfo) Name() string {
+	return i.name
+}
+
+func (i *DistributedObjectInfo) ServiceName() string {
+	return i.serviceName
+}
```
Comment (yuce) [inline types/distributed_objects_info.go:30]: No need for these anymore.

File: client.go [hunk 8]
```diff
@@ -200,6 +201,42 @@ func (c *Client) GetPNCounter(ctx context.Context, name string) (*PNCounter, err
 	return c.proxyManager.getPNCounter(ctx, name)
 }
 
+// GetDistributedObjectsInfo returns the information of all objects created cluster-wide.
+func (c *Client) GetDistributedObjectsInfo(ctx context.Context) ([]types.DistributedObjectInfo, error) {
+	if atomic.LoadInt32(&c.state) != ready {
+		return nil, hzerrors.ErrClientNotActive
+	}
+
```
Comment (yuce) [inline client.go:209]: Could you remove the blank lines in this function?

File: internal/proto/codec/builtin.go [hunk 9]
```diff
@@ -620,6 +620,24 @@ func DecodeListMultiFrameForDataContainsNullable(frameIterator *proto.ForwardFra
 	return result
 }
 
+func DecodeListMultiFrameForDistributedObjectInfo(frameIterator *proto.ForwardFrameIterator) []types.DistributedObjectInfo {
+	result := make([]types.DistributedObjectInfo, 0)
```
Comment (yuce) [inline internal/proto/codec/builtin.go:624]: Can be simplified to:
```
var result []types.DistributedObjectInfo
```

File: proxy.go [hunk 10]
```diff
@@ -53,6 +53,21 @@ const (
...
+		ServiceNameTopic,
+		ServiceNameList,
+		ServiceNameSet,
+		ServiceNamePNCounter:
+		return true
+	default:
+		return false
+	}
+}
+
```
Comment (yuce) [inline proxy.go:70]: We can remove this function.

File: types/distributed_objects_info.go [hunk 11]
```diff
@@ -0,0 +1,34 @@
...
+
+func NewDistributedObjectInfo(name string, serviceName string) DistributedObjectInfo {
+	return DistributedObjectInfo{name: name, serviceName: serviceName}
+}
```
Comment (yuce) [inline types/distributed_objects_info.go:34]: The user doesn't need this function, so it's better to move it to internal. How about `internal/util.go`?

**Review by yuce — CHANGES_REQUESTED** [review]
Thanks for the PR! Just a few comments to make it even better...

File: client.go (hunk 2 above)
Comment (enozcan) [reply client.go:222]: Isn't it still a good way of removing stale proxies?

File: types/distributed_objects_info.go (hunk 11 above)
Comment (enozcan) [reply types/distributed_objects_info.go:34]: I think no one needs it after exporting the fields. Removing it.

File: client.go (hunk 1 above)
Comment (enozcan) [reply client.go]: addressed in f2a4add

File: internal/it/util.go (hunk 5 above)
Comment (enozcan) [reply internal/it/util.go]: addressed in f2a4add

File: types/distributed_objects_info.go (hunk 6 above)
Comment (enozcan) [reply types/distributed_objects_info.go]: addressed in f2a4add

File: types/distributed_objects_info.go (hunk 7 above)
Comment (enozcan) [reply types/distributed_objects_info.go]: addressed in f2a4add

File: client.go (hunk 8 above)
Comment (enozcan) [reply client.go]: addressed in f2a4add

File: internal/proto/codec/builtin.go (hunk 9 above)
Comment (enozcan) [reply internal/proto/codec/builtin.go]: addressed in f2a4add

File: proxy.go (hunk 10 above)
Comment (enozcan) [reply proxy.go]: addressed in f2a4add

File: types/distributed_objects_info.go (hunk 11 above)
Comment (enozcan) [reply types/distributed_objects_info.go]: addressed in f2a4add

**Comment by enozcan** [comment]
@yuce addressed your comments in f2a4add except for purging the proxy cache - I left it unchanged to discuss further. Could you PTAL?

File: client.go (hunk 2 above)
Comment (yuce) [reply client.go:219]: I think purging stale proxies is a separate concern. As a user, I wouldn't expect `GetObjectsInfo` call to mutate any internal state.

Comment (yuce) [reply client.go:219]: Not sure about what to call the purge function though. If there's no immediate need from the Turbine team, maybe we can postpone adding it to another PR (unless you have a good name for that function) .

**Comment by yuce** [comment]
@enozcan Thanks for the update! I think this PR is good to be merged once the code related to proxy cache is removed.

File: client.go (hunk 2 above)
Comment (enozcan) [reply client.go]: Alright, fair enough.  Addressed in cd2887a

File: client.go (hunk 3 above)
Comment (enozcan) [reply client.go]: addressed in cd2887a

File: client.go (hunk 4 above)
Comment (enozcan) [reply client.go]: addressed in cd2887a

File: client.go (hunk 2 above)
Comment (enozcan) [reply client.go]: Apart from the context of this pr, could it be handled via DistributedObjectListeners in proxy manager - i.e. purge the cache on object destroy or something?

**Review by yuce — APPROVED** [review]
Thanks for the PR, it looks great!

File: client.go (hunk 2 above)
Comment (yuce) [reply client.go]: That can be an option, I can't think of a problem off the top of my head.

**Comment by enozcan** [comment]
Thanks for your review!

**Pull request merged**